plugins/logs/plugin.go

```diff
@@ -582,38 +617,66 @@ func (p *Plugin) loop() {
 			} else {
 				p.logger.Debug("Log upload queue was empty.")
 			}
-		}
-
-		var delay time.Duration
 
-		if err == nil {
-			min := float64(*p.config.Reporting.MinDelaySeconds)
-			max := float64(*p.config.Reporting.MaxDelaySeconds)
-			delay = time.Duration(((max - min) * rand.Float64()) + min)
-		} else {
-			delay = util.DefaultBackoff(float64(minRetryDelay), float64(*p.config.Reporting.MaxDelaySeconds), retry)
+			if *p.config.Reporting.Trigger == plugins.TriggerManual {
+				close(mtDone)
+				wait = true
+			}
 		}
 
-		if p.config.Service != "" {
-			p.logger.Debug("Waiting %v before next upload/retry.", delay)
-		}
+		var waitC chan chan struct{}
 
-		timer := time.NewTimer(delay)
+		switch *p.config.Reporting.Trigger {
+		case plugins.TriggerPeriodic:
+			var delay time.Duration
 
-		select {
-		case <-timer.C:
-			if err != nil {
-				retry++
+			if err == nil {
+				min := float64(*p.config.Reporting.MinDelaySeconds)
+				max := float64(*p.config.Reporting.MaxDelaySeconds)
+				delay = time.Duration(((max - min) * rand.Float64()) + min)
 			} else {
-				retry = 0
+				delay = util.DefaultBackoff(float64(minRetryDelay), float64(*p.config.Reporting.MaxDelaySeconds), retry)
 			}
-		case update := <-p.reconfig:
-			p.reconfigure(update.config)
-			update.done <- struct{}{}
+
+			if p.config.Service != "" {
+				p.logger.Debug("Waiting %v before next upload/retry.", delay)
+			}
+
+			waitC = make(chan chan struct{}, 1)
+
+			go func() {
+				select {
+				case <-time.After(delay):
+					if err != nil {
+						retry++
+					} else {
+						retry = 0
+					}
+
+					if ctx.Err() != nil {
+						return
+					}
+
+					close(waitC)
+				case <-ctx.Done():
+					return
+				}
+			}()
+		case plugins.TriggerManual:
+			p.logger.Debug("Waiting for manual trigger before next upload/retry.")
+			waitC = p.trigger
+		}
+
+		select {
+		case mtDone = <-waitC:
+			wait = false
 		case done := <-p.stop:
 			cancel()
 			done <- struct{}{}
 			return
+		case update := <-p.reconfig:
+			p.reconfigure(update.config)
+			update.done <- struct{}{}
 		}
 	}
```

Comment:
❓ Is the logic happening in the different plugins for periodic vs triggered runs similar enough to be spun out into a little helper package?